### ISSUE TITLE
SIMD is broken

### ISSUE THREAD

**Issue by maxim-saplin** [dart-lang/sdk#53662]
**STEPS:**

1. Copy the file: https://github.com/maxim-saplin/mandelbrot/blob/main/_optimized/mandelbrot.dart
2. RUN it in VM `dart mandelbrot.dart ` and remember the results (execution time and check sum)
3. Build it with AoT `dart compile exe mandelbrot.dart`, run and remember the result 

**EXPECTED**:
- VM and AoT versions have similar performance (i.e. +/- 10%)
- VM and AoT have correct result (+/- 1% from 78513425) 

**ACTUAL**:
- AoT mode is ~40x slower on both Apple Silicon and Intel
- AoT with Intel has wrong sum 87667671

// M1 Pro, sum 78513692
// dart mandelbrot.dart - Avg: 93.4ms, StdDev: 1.6119%
// dart compile exe mandelbrot.dart - Avg: 4038.5ms, StdDev: 0.6437%
// Intel
// dart mandelbrot.dart - !! sum 87667671 Avg: 162.9ms, StdDev: 7.5598%
// dart compile exe mandelbrot.dart - sum 78513692, Avg: 8806.0ms, StdDev: 4.4871%

**ENVIRONMENTS:**

- macOS 13.6, Dart SDK version: 3.1.2, MacBook Pro with M1 Pro CPU
- Ubuntu 22.04.3 LTS, 64 bit, Dart SDK version: 3.1.0, Intel Core i5-8257U @ 1.4GHz x 2, VMWare Workstation Player 17.0.1

**Comment by lrhn** [comment]
I notice that the first three rounds have a different checksum on JIT too, suggesting that the code is susceptible to differences in rounding of intermediate values.

That's a little worrisome, since the code appears like it should be deterministic. Every operation is a SIMD 32-bit floating point addition or multiplication, or a comparison.

Getting different results anyway suggests either not using the same SIMD operations before and after JIT optimization, having different rounding behavior, not using SIMD before optimizing at all, or just something being bugged.

**Comment by maxim-saplin** [comment]
Here're a SO thread, few guys checked assembly and told it is horrible (check comments): https://stackoverflow.com/questions/77201568/dart-simd-extensions-int32x4-float32x4-going-crazy-slow-in-aot-different-re

Regarding the the correctness of the results, what is not OK is the result on Intel, it is way to far from what I would expect (10%). JIC, I've tested same problem with 8 different languages (22 configs) they all differ slightly in precision and check sum: https://github.com/maxim-saplin/mandelbrot/tree/main

And performance in AoT is another problem, relevant for both Intel and Apple Silicon.

**Comment by lrhn** [comment]
The difference in sum is around ~1M, which is on the order of one iteration per pixel. That is a very big difference, bigger than what (I think) can be explained by rounding. Every iteration does a squaring, which should quench low-bit rather than amplify them. (If you multiply two 24 bit integers, then keep the top 24 bits of the result, that should be less affected by low-bit changes than addition.)

**Comment by a-siva** [comment]
//cc @rmacnak-google @alexmarkov

**Comment by a-siva** [comment]
The correctness issue has been fixed, the performance issue is still open.
[https://github.com/dart-lang/sdk/issues/31487](https://github.com/dart-lang/sdk/issues/31487) is tracking the need for const constructors so Flutter code could use SIMD.

**Comment by gyrdym** [comment]
Hi everyone,

Are there any updates on this issue?

I'm experiencing the same problem: my library, which heavily utilizes SIMD, is extremely slow in AoT mode.

**Comment by maks** [comment]
Given https://github.com/dart-lang/sdk/commit/e0710a5cc61a52b73b07a616e2c611ef79b57a60 is in merge and now even in stable, does that now fix the perf part of this issue and https://github.com/dart-lang/sdk/issues/31487 ?

**Comment by mraleph** [comment]
I don't think this performance issue was related to https://github.com/dart-lang/sdk/commit/e0710a5cc61a52b73b07a616e2c611ef79b57a60 - though it is the same in nature. Likely some missing specialization in AOT version of call specializer.

**Comment by maks** [comment]
Thanks for looking at this again @mraleph ! 🙏🏻 

Sorry I'm a bit confused now, does that mean that the performance poor performance of Float32x4 multiplication or addition in AOT is still an unresolved issue?

I got here while I was doing some research into using Dart SIMD for an upcoming project via the comment in this other repo: https://github.com/gyrdym/ml_linalg/issues/175#issuecomment-2282903755 which reported your commit that went into Dart 3.5 as having fixed the performance issue they were seeing with using Float32 in AOT builds.

**Comment by modulovalue** [comment]
https://dart-review.googlesource.com/c/sdk/+/497000
